aggregator/src/aggregator/queue.rs

```diff
@@ -0,0 +1,567 @@
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Arc, Mutex as StdMutex,
+};
+
+use tokio::{
+    sync::{
+        mpsc,
+        oneshot::{self},
+        Notify, OwnedSemaphorePermit, Semaphore,
+    },
+    task::JoinHandle,
+};
+use tracing::warn;
+use trillium::{Conn, Handler};
+use trillium_macros::Handler;
+
+use super::Error;
+
+type Stack = Arc<StdMutex<Vec<Dispatcher>>>;
+
+/// A queue that services requests in a LIFO manner, i.e. the most recent request is serviced first.
+/// It bounds the total number of waiting requests, and the number of requests that can be run
+/// concurrently. This is useful for adding backpressure and preventing the process from being
+/// overwhelmed.
+///
+/// See https://encore.dev/blog/queueing for a rationale of LIFO request queuing.
+///
+/// Note the actual execution order of requests is not perfectly LIFO if the concurrency is >1,
+/// because the order that request futures are scheduled and executed in is essentially
+/// non-deterministic.
+#[derive(Debug)]
+pub struct LIFORequestQueue {
+    /// Maximum size of the queue.
+    depth: usize,
+
+    /// LIFO data structure.
+    stack: Stack,
+
+    /// Notifies the dispatch task to wake up when a new ticket arrives in an empty stack.
+    notifier: Arc<Notify>,
+
+    /// Controls concurrency of requests.
+    _semaphore: Arc<Semaphore>,
+
+    /// Alerts tickets when they're ready.
+    dispatcher: JoinHandle<()>,
+
+    /// Generates unique ticket IDs, used to identify tickets in the queue.
+    id_generator: TicketIdGenerator,
+
+    /// Receives notifications when a request has given up on waiting in the queue.
+    cancel_tx: mpsc::UnboundedSender<TicketId>,
+
+    /// Services cancel notifications by directly removing cancelled tickets from the queue. This
+    /// prevents cancelled tickets from taking up space in the queue and starving new requests.
+    canceller: JoinHandle<()>,
+}
+
+impl LIFORequestQueue {
+    /// Creates a new [`Self`].
+    ///
+    /// `concurrency` and `depth` must be greater than 0, and `depth` must be greater than or
+    /// equal to `concurrency`.
+    pub fn new(concurrency: usize, depth: usize) -> Result<Self, Error> {
+        if concurrency < 1 {
+            return Err(Error::InvalidConfiguration(
+                "concurrency must be greater than 0",
+            ));
+        } else if depth < 1 {
+            return Err(Error::InvalidConfiguration(
+                "depth must be greater than zero",
+            ));
+        } else if concurrency > depth {
+            // We enforce this property otherwise it leads to strange intermittent behavior, i.e.
```

Comment:
The behavior I'm describing is intermittent, because it depends on how the dispatcher task is scheduled.

Suppose we have concurrency 10 and depth 1. Suppose 5 requests come in and are scheduled before the dispatcher task. 1 request will be serviced and the remaining 4 requests will be rejected, because there is no queue space and the dispatcher hasn't run yet.

Part of the problem is that requests are queued first before being run, i.e. concurrency slots are not immediately taken.

I think I may be able to better hide this property by shuffling around variables. I can make the "true" depth of the queue `depth + concurrency`, i.e. turn the `if stack.len() < self.depth` check into `if stack.len() < self.depth + self.concurrency`.

Reply:
> I think I may be able to better hide this property by shuffling around variables. I can make the "true" depth of the queue depth + concurrency, i.e. turn the if stack.len() < self.depth check into if stack.len() < self.depth + self.concurrency.

Eh, this isn't great either. It means we'll always queue `self.concurrency` requests at a time, at minimum, which confounds the notion of having, e.g., depth 0 and concurrency 10. (We'd always still queue 10 requests).

Maybe `.acquire()` should `semaphore.try_acquire()`, and return immediately with a permit if it succeeds, otherwise return a ticket to wait in line.

Reply:
> Maybe .acquire() should semaphore.try_acquire(), and return immediately with a permit if it succeeds, otherwise return a ticket to wait in line.

This ended up pretty nice IMO. `.acquire()` now will return a `Ticket` which can be either an immediately ready semaphore permit, or a oneshot channel to receive one at a later time. Quickcheck tests with queue depth zero are passing now.

Reply:
I've pushed it. n.b. it currently has a transient deadlock in `quickcheck_lifo_full()` I'm currently debugging.

I've not reviewed your proposal in depth yet, but

> this also simplifies the "concurrency" handling from shared-semaphore/channels/shared-state-via-mutexes/notifier to just a private semaphore & channels.

is an attractive proposition. The fewer concurrency primitives, the better.

Reply:
I've gone with your approach @branlwyd, the algorithm proves remarkably tight. https://github.com/divviup/janus/pull/3369/files#diff-2390b77ede40036c299d10421fb7b47340e1f6c38766f2e4cffd562fb5e8396bR77.

Thanks for the suggestion!